andela/ah-codeblooded-frontend
# #162192206 Intercept and handle network errors

**What does this PR do?**

This PR handles network errors application-wide. This is to prevent developers from handling them everywhere that they make API calls.

**Description of Task to be completed?**

Network errors should be handled at one point to prevent repetition of code. Since the entire application uses a single `api` module that wraps `axios` to make http requests, all network errors should be handled by it. A modal should be used to indicate network errors. All areas of the application that have handled network errors should be refactored to be in line with this feature.

**How should this be manually tested?**

This can be tested by following these steps:
1. Open this PRs review app.
2. Click to open any article on the homepage. Allow it to load.
3. Go to your browser dev tools and simulate that the page is offline by checking Offline on the network tab.
4. Wait for 30 seconds and a modal will appear indicating that you're offline. The network error is automatically detected when the application tries to update article reactions and fails.
5. Go back to your browser dev tools disable offline mode on the network tab.
6. Wait for 30 seconds and the network error modal will disappear indicating that there is a connection.

**Any background context you want to provide?**

This solution is making use of `axios` interceptors, https://github.com/axios/axios#interceptors.

**What are the relevant pivotal tracker stories?**

[#162192206](https://www.pivotaltracker.com/story/show/162192206)

**Screenshots**

Simulating network error on browser dev tools.
<img width="436" alt="screen shot 2018-11-27 at 23 52 22" src="https://user-images.githubusercontent.com/8180548/49110826-93e85980-f29f-11e8-9f1c-b26b5204d3d7.png">

Viewing an article
<img width="875" alt="screen shot 2018-11-29 at 07 09 59" src="https://user-images.githubusercontent.com/8180548/49198902-6fc37000-f3a6-11e8-9220-3ad28275782d.png">

Network error occurs while viewing
<img width="872" alt="screen shot 2018-11-29 at 07 10 39" src="https://user-images.githubusercontent.com/8180548/49198913-7e118c00-f3a6-11e8-89af-d02c8e6fb87a.png">

**Questions**
Anyone that has successfully tested a materialize modal can give ideas on how we can test this.
